Spigot-Server-Patches/0424-Added-EndGatewayCreationEvent.patch

```diff
@@ -0,0 +1,60 @@
+From ba3030b5ea03ff18b4f2a86858e571dcb35136d5 Mon Sep 17 00:00:00 2001
+From: JOO200 <git@johanneszangl.de>
+Date: Tue, 19 Mar 2019 20:47:13 +0100
+Subject: [PATCH] Added EndGatewayCreationEvent
+
+
+diff --git a/src/main/java/net/minecraft/server/EnderDragonBattle.java b/src/main/java/net/minecraft/server/EnderDragonBattle.java
+index 09eabf123..87e6b63e5 100644
+--- a/src/main/java/net/minecraft/server/EnderDragonBattle.java
++++ b/src/main/java/net/minecraft/server/EnderDragonBattle.java
+@@ -1,5 +1,6 @@
+ package net.minecraft.server;
+ 
++import com.destroystokyo.paper.event.block.EndGatewayCreationEvent;
+ import com.google.common.collect.ContiguousSet;
+ import com.google.common.collect.DiscreteDomain;
+ import com.google.common.collect.Lists;
+@@ -15,14 +16,16 @@ import java.util.function.Predicate;
+ import javax.annotation.Nullable;
+ import org.apache.logging.log4j.LogManager;
+ import org.apache.logging.log4j.Logger;
++import org.bukkit.Bukkit;
++import org.bukkit.Location;
+ 
+ public class EnderDragonBattle {
+ 
+     private static final Logger a = LogManager.getLogger();
+     private static final Predicate<Entity> b = IEntitySelector.a.and(IEntitySelector.a(0.0D, 128.0D, 0.0D, 192.0D));
+     public final BossBattleServer bossBattle;
+-    private final WorldServer d;
+-    private final List<Integer> e;
++    private final WorldServer d; public final WorldServer getWorldServer() { return this.d; } // Paper - OBFHELPER
++    private final List<Integer> e; public final List<Integer> getSpawnableGateways() { return this.e; } // Paper - OBFHELPER
+     private final ShapeDetector f;
+     private int g;
+     private int h;
+@@ -385,10 +388,18 @@ public class EnderDragonBattle {
+ 
+     private void m() {
+         if (!this.e.isEmpty()) {
+-            int i = (Integer) this.e.remove(this.e.size() - 1);
++            // Paper start - implement EndGatewayCreationEvent
++            int i = (Integer) this.getSpawnableGateways().get(this.getSpawnableGateways().size() - 1);
+             int j = (int) (96.0D * Math.cos(2.0D * (-3.141592653589793D + 0.15707963267948966D * (double) i)));
+             int k = (int) (96.0D * Math.sin(2.0D * (-3.141592653589793D + 0.15707963267948966D * (double) i)));
+-
++            Location loc = MCUtil.toLocation(this.getWorldServer(), j, 75, k);
++            EndGatewayCreationEvent event = new EndGatewayCreationEvent(loc.getBlock());
++            Bukkit.getPluginManager().callEvent(event);
++            if (event.isCancelled()) {
```

Comment:
```suggestion
+            if (!event.callEvent()) {
```
and remove the line above and the import

Reply:
in general, fully qualified imports are preferred as they reduce the size of the diff, so I would do that for the event and location.